### ISSUE TITLE
Nonblocking way working with db

### ISSUE THREAD

**Issue by EgorSidorov** [oatpp/example-postgresql#1]
Good day!

Great framework and very useful example!
I see that postgreSQL interacts in blocking mode when using coroutin in oatpp. I would also like to look at the possibility of working with the database in non-blocking mode(example: https://www.postgresql.org/docs/9.0/libpq-async.html), since, I believe, the main advantage of oatpp is the use of coroutin.

**Comment by lganzzzo** [comment]
Hello @EgorSidorov ,

Thanks for the question and for your feedback!

The recommended way to use Async APIs and databases is as follows - **see the diagram.**
The point is that it's not a good idea to let all the incoming connections straight to your Database. DB connections are very costly (on the DB instance side).
Also, sometimes it's a good idea to pool connections on the gateway side and do not let all the incoming connections straight into your system.


![async-to-db](https://user-images.githubusercontent.com/5940640/81574388-69646080-93ae-11ea-9c8c-d8986a4007a1.png)


In case you have a monolith application, you can just merge your distributed services in one using the oatpp-monolithization approach - https://oatpp.io/docs/monolithization/

Thus you can keep oatpp-coroutines simple (by letting them do a large number of simple tasks),
and you can keep your business code clean by utilizing oatpp simple API with all the Request-Mapping and Object-Mapping framework.

*Also, if you don't want to build multiple services for your case, you can then use 
`oatpp::async::Lock` and `oatpp::async::CoroutineWaitList` to delegate the processing of DB tasks from coroutines to a thread pool and then notify coroutine upon task completion.*

Please let me know if you have more questions.

Regards,
Leonid

**Comment by EgorSidorov** [comment]
Thank you for answer!

This is an interesting approach to working with the database, I did not think about creating a microservice for working with the database in synchronous mode. But still, the approach with the thread pool in the application looks simpler in terms of the number of microservices. 
Can you please give me some examples of using oatpp::async::CoroutineWaitList for waiting of std::thread? I only have an understanding that I can run thread and wait std::atomic<bool> condition with repeat() coroutine until condition != true.

**Comment by EgorSidorov** [comment]
I understood, I should use oatpp::async::Lock try_lock() and waitAsync() from coroutine before executing sql and unlock() from std::thread when getting sql result. Thanks for the advice.

**Comment by lganzzzo** [comment]
Hello @EgorSidorov ,

Not really, I'll prepare the example code-snippet for you later today.

- The coroutine should place an SQL task in the thread pool.
- Once the task is queued it should put the corresponding coroutine on a waitlist.
- Once the task is finished it should notify the coroutine to go back to processing.

Regards,
Leonid

**Comment by lganzzzo** [comment]
Hey @EgorSidorov ,

Here is the full working example of how-to make coroutine wait for the thread.
Make sure to use a **thread-pool** for your threads! 

```cpp
#include "oatpp/core/async/CoroutineWaitList.hpp"
#include "oatpp/core/async/Executor.hpp"
#include "oatpp/core/utils/ConversionUtils.hpp"
#include "oatpp/core/Types.hpp"

typedef oatpp::async::CoroutineWaitList CoroutineWaitList;
typedef oatpp::async::Action Action;

class QueryTask : private CoroutineWaitList::Listener {
private:
  // Is called from insede the async::Processor when coroutine was added on a waitlist.
  void onNewItem(CoroutineWaitList& list) override {
    if(m_finished) {
      list.notifyFirst();
    }
  }
private:
  oatpp::String m_query;
  std::atomic<bool> m_finished;
  CoroutineWaitList m_list;
public:

  QueryTask(const oatpp::String& query)
    : m_query(query)
    , m_finished(false)
  {
    m_list.setListener(this);
  }

  Action runTaskAsyncInline(Action&& nextAction) {

    if(m_finished) {
      return std::forward<Action>(nextAction);
    }

    std::thread t([this]{ // <-- TODO Pool these threads !!!

      OATPP_LOGD("Query", "Executing query: %s ...", m_query->c_str());
      std::this_thread::sleep_for(std::chrono::seconds(3)); // wait for 3 sec

      m_finished = true;
      m_list.notifyFirst();
    });

    t.detach();

    // Action to tell the processor to put the executing coroutine
    // on a wait list so that processor wont spin while this coroutine
    // is waiting for the thread to finish.
    return Action::createWaitListAction(&m_list);

  }

};

class QueryCoroutine : public oatpp::async::Coroutine<QueryCoroutine> {
private:
  QueryTask m_task;
public:

  QueryCoroutine(const oatpp::String& query)
    : m_task(query)
  {}

  Action act() override {
    return m_task.runTaskAsyncInline(yieldTo(&QueryCoroutine::onTaskDone));
  }

  Action onTaskDone() {
    OATPP_LOGD("QueryCoroutine", "task is done!");
    return finish();
  }

};

void run() {

  // Executor with one thread
  oatpp::async::Executor executor(1);

  // Populate 100 coroutines
  for (v_int32 i = 0; i < 100; i++) {
    executor.execute<QueryCoroutine>("SELECT FROM BLAH... - query_" + oatpp::utils::conversion::int32ToStr(i));
  }

  executor.waitTasksFinished();
  executor.stop();
  executor.join();

}

int main() {
  oatpp::base::Environment::init();
  run();
  oatpp::base::Environment::destroy();
  return 0;
}
```

Please let me know if you have any questions!

Best Regards,
Leonid

**Comment by EgorSidorov** [comment]
Thank you very much!

**Comment by lganzzzo** [comment]
Hey @EgorSidorov ,

Just in case you decide to go with multiple services - 
I've created a small project [funnel-proxy](https://github.com/oatpp/funnel-proxy) for similar to your cases. Feel free to fork/contribute.

Regards,
Leonid
